File: epi_judge_java/epi/AdvanceByOffsets.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.List;

public class AdvanceByOffsets {

    @EpiTest(testDataFile = "advance_by_offsets.tsv")
    public static boolean canReachEnd(List<Integer> maxAdvanceSteps) {
        var maxSoFar = 0;
        for (int i = 0; i < maxAdvanceSteps.size(); i++) {
            if (maxSoFar < i) return false;
            maxSoFar = Math.max(maxSoFar, i + maxAdvanceSteps.get(i));
            if (maxSoFar >= maxAdvanceSteps.size()) break;
        }
        return true;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "AdvanceByOffsets.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_python/advance_by_offsets.py
from typing import List

from test_framework import generic_test


def can_reach_end(arr: List[int]) -> bool:
    next_furthest = 0
    i = 0
    last_index = len(arr) - 1

    while i <= next_furthest < last_index:
        next_furthest = max(next_furthest, i + arr[i])
        i += 1

    return next_furthest >= last_index


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('advance_by_offsets.py',
                                       'advance_by_offsets.tsv',
                                       can_reach_end))
